lib/bridge/IrcHandler.js

```diff
@@ -176,6 +185,86 @@ IrcHandler.prototype.onPrivateMessage = Promise.coroutine(function*(req, server,
     yield this.ircBridge.sendMatrixAction(pmRoom, virtualMatrixUser, mxAction, req);
 });
 
+IrcHandler.prototype._serviceTopicQueue = Promise.coroutine(function*(item) {
+    let promises = item.entries.map((entry) => {
+        if (entry.matrix.topic === item.topic) {
+            item.req.log.info(
+                "Topic of %s already set to '%s'",
+                entry.matrix.getId(),
+                item.topic
+            );
+            return Promise.resolve();
+        }
+        console.log('setRoomTopic');
+        return this.ircBridge.getAppServiceBridge().getIntent().setRoomTopic(
+            entry.matrix.getId(),
+            item.topic
+        ).then(
+            () => {
+                console.log('upsertRoomStoreentry');
+                entry.matrix.topic = item.topic;
+                return this.ircBridge.getStore().upsertRoomStoreEntry(entry);
+            },
+            (err) => {
+                item.req.log.error(`Error storing room ${entry.matrix.getId()} (${err.message})`);
+            }
+        );
+    });
+    try {
+        yield Promise.all(promises);
+        item.req.log.info(
+            "Set topics: ",
+            item.entries.map(
+                (entry) => `id = ${entry.matrix.getId()}, topic = '${entry.matrix.topic}'`
+            ).join('\n')
+        );
+    }
+    catch (err) {
+        req.log.error('Failed to set topic(s) ${err.message}');
+    }
+});
+
+/**
+ * Called when the AS receives an IRC topic event.
+ * @param {IrcServer} server : The sending IRC server.
+ * @param {IrcUser} fromUser : The sender.
+ * @param {string} channel : The target channel.
+ * @param {Object} action : The IRC action performed.
+ * @return {Promise} which is resolved/rejected when the request finishes.
+ */
+IrcHandler.prototype.onTopic = Promise.coroutine(function*(req, server, fromUser,
+                                                    channel, action) {
+    req.log.info("onTopic: %s from=%s to=%s action=%s",
+        server.domain, fromUser, channel, JSON.stringify(action).substring(0, 80)
+    );
+
+    let topic = action.text;
+
+    // Only bridge topics for rooms created by the bridge, via !join or an alias
+    let origins = ["join", "alias"];
+    let entries = yield this.ircBridge.getStore().getMappingsForChannelByOrigin(
+        server, channel, origins
+    );
+    if (entries.length === 0) {
+        req.log.info(
+            "No mapped matrix rooms for IRC channel %s with origin = [%s]",
+            channel,
+            origins
+        );
+        return;
+    }
+
+    req.log.info(
+        "New topic in %s - bot queing to set topic in %s",
+        channel,
+        entries.map((e) => e.matrix.getId())
+    );
+    yield this.topicQueue.enqueue(
+        server.domain + " " + channel + " " + topic,
```

Comment:
Oh, you're right, I'm misreading this.

So we have a single global queue for every topic change in the entire bridge? Can we map this to be per-channel please to avoid head of line blocking?